scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -279,9 +279,121 @@
+            progress-aggregator-trigger-chan
+            (do
+              (when (seq instance-id->progress-state)
+                (log/debug "Forwarding" (count instance-id->progress-state) "progress states to the transactor")
+                (histograms/update! progress-aggregator-pending-states (count instance-id->progress-state))
+                (async/>! progress-state-chan instance-id->progress-state))
+              (recur {}))
+
+            progress-aggregator-chan
+            (let [{:keys [instance-id]} data]
+              (meters/mark! progress-aggregator-message-rate)
+              (if (or (< (count instance-id->progress-state) pending-progress-threshold)
+                      (contains? instance-id->progress-state instance-id))
+                (let [progress-state (select-keys data [:progress-message :progress-percent])
+                      instance-id->progress-state' (assoc instance-id->progress-state instance-id progress-state)]
+                  (recur instance-id->progress-state'))
+                (do
+                  (meters/mark! progress-aggregator-drop-rate)
+                  (counters/inc! progress-aggregator-drop-count)
+                  (log/debug "Dropping" data "as threshold has been reached")
+                  (recur instance-id->progress-state))))
+
+            query-state-chan
+            (let [{:keys [mode response-chan]} data]
+              (condp = mode
+                :query
+                (do
+                  (async/>! response-chan instance-id->progress-state)
+                  (recur instance-id->progress-state))
+
+                :exit
+                (do
+                  (when response-chan
+                    (async/>! response-chan :exited))
+                  (log/info "Exiting progress update aggregator"))))))))
+    {:progress-aggregator-chan progress-aggregator-chan
+     :query-state-chan query-state-chan}))
+
+(histograms/defhistogram [cook-mesos scheduler progress-updater-pending-states])
+(meters/defmeter [cook-mesos scheduler progress-updater-publish-rate])
+(timers/deftimer [cook-mesos scheduler progress-updater-publish-duration])
+(meters/defmeter [cook-mesos scheduler progress-updater-tx-rate])
+(timers/deftimer [cook-mesos scheduler progress-updater-tx-duration])
+
+(defn progress-update-transactor
+  "Launches a long running go block that triggers transacting the latest aggregated instance-id->progress-state to
+   datomic whenever there is a message on the progress-updater-trigger-chan.
+   No more than batch-size facts are updated in individual datomic transactions.
+   It returns a progress-state-chan which can be used to send messages about the latest instance-id->progress-state.
+   If no such message is on the progress-state-chan, then no datomic interactions occur."
+  [progress-updater-trigger-chan batch-size conn]
+  (log/info "Starting progress update transactor")
+  (let [progress-state-chan (async/chan (async/sliding-buffer 1))]
+    (async/go
```

Comment:
I think you can replace this go and the loop with https://github.com/twosigma/Cook/blob/master/scheduler/src/cook/mesos/util.clj#L523 which handles waiting on the trigger channel and then closing the response channel.